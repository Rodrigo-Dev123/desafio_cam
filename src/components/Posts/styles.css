.post-content {
    margin: 0 auto;
    padding: 10px;
    display: flex;
    align-items: center;
    width: max-content;
    gap: 5px
}

.container {
    min-height: 100vh;
    position: relative;
}

.post {
    box-shadow: 0px 0px 10px 5px rgba(0, 0, 0, 0.315);
    background-color: white;
    display: flex;
    flex-direction: column;
}

.post img {
    padding: 10px;
}

.btn-delete-post {
    border: none;
    background-color: transparent;
    cursor: pointer;
}

.post:hover {
    transform: scale(1.05);
    transition: ease-in-out, 300ms;
}

.icon-delete {
    display: block;
    width: 35px;
    height: 35px;
}
.icon-delete:hover {
    color: red;
}