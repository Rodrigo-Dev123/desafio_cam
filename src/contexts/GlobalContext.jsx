import P from 'prop-types';
import { createContext, useState, useEffect } from "react";

export const GlobalContext = createContext();

const GlobalProvider = ({ children }) => {
    const [items, setItems] = useState([]);
    const [itemsPerPage] = useState(10);
    const [currentPage, setCurrentPage] = useState(0);

    const [modal, setModal] = useState(false);
    const [postModal, setPostModal] = useState(null);

    const startIndex = currentPage * itemsPerPage;
    const endIndex = startIndex + itemsPerPage;
    const pages = Math.ceil(items.length / itemsPerPage);
    const currentItems = items.slice(startIndex, endIndex);

    const removeItem = (id) => {
        const updatedItems = JSON.parse(localStorage.getItem('items')).filter(item => {
            return item.id !== id;
        })

        localStorage.setItem('items', JSON.stringify(updatedItems))
        setItems(JSON.parse(localStorage.getItem('items')));
        alert(`Foto ${id} apagada com sucesso! ｡◕‿◕｡`);
    }

    useEffect(() => {
        if (!localStorage.getItem('items')) {
            const fetchData = async () => {
                console.log('fetch execultou!');
                const result = await fetch('https://jsonplaceholder.typicode.com/photos')
                    .then(response => response.json())
                    .then(data => data)

                localStorage.setItem('items', JSON.stringify(result));
                setItems(result);
            }

            fetchData();
            return;
        }
        setItems(JSON.parse(localStorage.getItem('items')));
    }, []);

    return (
        <GlobalContext.Provider value={{
            currentItems, items, setItems, itemsPerPage,
            currentPage, setCurrentPage, modal, setModal,
            startIndex, endIndex, pages, removeItem,
            postModal, setPostModal
        }}>
            {children}
        </GlobalContext.Provider>
    )
}

GlobalProvider.propTypes = {
    children: P.node.isRequired,
};

export default GlobalProvider;
