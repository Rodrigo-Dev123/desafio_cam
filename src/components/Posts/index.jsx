import P from 'prop-types';
import './styles.css';
import { useContext } from "react";
import ButtonModal from "../buttonModal";
import Modal from "../Modal";
import { GlobalContext } from "../../contexts/GlobalContext";
import Pagination from "../Pagination";
import { RiDeleteBin6Line } from "react-icons/ri";

export default function Posts({ post = [] }) {
    const { removeItem, currentItems, } = useContext(GlobalContext);


    return (
        <section className='container'>
            <div className="App">
                {currentItems.map((post) => {
                    return (
                        <div key={post.id} className="post">
                            <img src={post.url} alt={post.title}></img>
                            <h3>Foto {post.id}</h3>
                            <div className='post-content'>
                                <ButtonModal id={post.id}/>
                                <button className='btn-delete-post' onClick={() => removeItem(post.id)}>
                                    <RiDeleteBin6Line
                                        className='icon-delete'
                                    />
                                </button>
                            </div>
                        </div>
                    );
                })}
                <Modal />
            </div>


            <Pagination />
        </section>
    );
};

Posts.propTypes = {
    posts: P.array,
};