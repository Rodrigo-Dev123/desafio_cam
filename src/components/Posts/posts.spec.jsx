import { render, screen } from '@testing-library/react';
import { GlobalContext } from '../../contexts/GlobalContext';
import Posts from '.';

const contextValue = {
  removeItem: jest.fn(),
  currentItems: [
    {
      "albumId": 1,
      "id": 1,
      "title": "accusamus beatae ad facilis cum similique qui sunt",
      "url": "https://via.placeholder.com/600/92c952",
      "thumbnailUrl": "https://via.placeholder.com/150/92c952"
    },
    {
      "albumId": 1,
      "id": 2,
      "title": "reprehenderit est deserunt velit ipsam",
      "url": "https://via.placeholder.com/600/771796",
      "thumbnailUrl": "https://via.placeholder.com/150/771796"
    },
    {
      "albumId": 1,
      "id": 3,
      "title": "officia porro iure quia iusto qui ipsa ut modi",
      "url": "https://via.placeholder.com/600/24f355",
      "thumbnailUrl": "https://via.placeholder.com/150/24f355"
    },
  ],
};

describe('<Posts />', () => {
  it('Should render posts', () => {
    const { debug } = render(
      <GlobalContext.Provider value={contextValue}>
        <Posts />
      </GlobalContext.Provider>
    );
    debug();

    expect(screen.getAllByRole('heading', { name: /Foto/i })).toHaveLength(contextValue.currentItems.length);
  });

  it('Shouls math snapshot', () => {
    const { container } = render(
      <GlobalContext.Provider value={contextValue}>
        <Posts />
      </GlobalContext.Provider>
    );
  
    expect(container.firstChild).toMatchSnapshot();
  });
});