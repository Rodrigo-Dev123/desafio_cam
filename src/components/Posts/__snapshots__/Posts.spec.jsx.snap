// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> Shouls math snapshot 1`] = `
<section
  class="container"
>
  <div
    class="App"
  >
    <div
      class="post"
    >
      <img
        alt="accusamus beatae ad facilis cum similique qui sunt"
        src="https://via.placeholder.com/600/92c952"
      />
      <h3>
        Foto 
        1
      </h3>
      <div
        class="post-content"
      >
        <input
          id="1"
          type="button"
          value="Ver Mais"
        />
        <button
          class="btn-delete-post"
        >
          <svg
            class="icon-delete"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="post"
    >
      <img
        alt="reprehenderit est deserunt velit ipsam"
        src="https://via.placeholder.com/600/771796"
      />
      <h3>
        Foto 
        2
      </h3>
      <div
        class="post-content"
      >
        <input
          id="2"
          type="button"
          value="Ver Mais"
        />
        <button
          class="btn-delete-post"
        >
          <svg
            class="icon-delete"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="post"
    >
      <img
        alt="officia porro iure quia iusto qui ipsa ut modi"
        src="https://via.placeholder.com/600/24f355"
      />
      <h3>
        Foto 
        3
      </h3>
      <div
        class="post-content"
      >
        <input
          id="3"
          type="button"
          value="Ver Mais"
        />
        <button
          class="btn-delete-post"
        >
          <svg
            class="icon-delete"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="pagination"
  >
    <div
      class="container-button"
    >
      <div>
        <button
          class="btn-page"
          value="0"
        >
          1
        </button>
      </div>
    </div>
  </div>
</section>
`;
